libkatana_python_native/include/katana/python/NumbaSupport.h

```diff
@@ -0,0 +1,146 @@
+
+  static remove_cvref_t<Return> call(
+      ConstMemberFunction<Cls, Return, Args...>* func, const Cls* self,
+      remove_cvref_t<Args>... args) {
+    return (self->*(func->member_func))(args...);
+  }
+};
+
+template <typename... Args>
+class DefWithNumbaImpl {
+  template <
+      typename Cls, typename Return, typename Func, typename Caller,
+      typename... Extra>
+  void def_class_method(
+      pybind11::class_<Cls>& cls, const char* name, Func f, Caller* caller,
+      const Extra&... extra) const {
+    cls.def(name, f, extra...);
+    auto numba_support =
+        pybind11::module_::import("katana.native_interfacing.numba_support");
+    numba_support.attr("register_method")(
+        cls, cls.attr(name), (uintptr_t)&Caller::call, (uintptr_t)caller,
+        PythonTypeTraits<remove_cvref_t<Return>>::ctypes_type(),
+        PythonTypeTraits<remove_cvref_t<Args>>::ctypes_type()...);
+  }
+
+  template <typename Return, typename... Extra>
+  void def_func(
+      pybind11::module_& m, const char* name, Return (*f)(Args...),
+      const Extra&... extra) const {
+    m.def(name, f, extra...);
+    auto func = m.attr(name);
+    auto numba_support =
+        pybind11::module_::import("katana.native_interfacing.numba_support");
+    numba_support.attr("register_function")(
+        func, (uintptr_t)f, 0,
+        PythonTypeTraits<remove_cvref_t<Return>>::ctypes_type(),
+        PythonTypeTraits<remove_cvref_t<Args>>::ctypes_type()...);
+  }
+
+public:
+  // TODO(amp): The generated wrappers used from numba code are created per
+  //  *signature*, not per function. So two functions will the same overall
+  //  signature (including containing class) will use the same wrapper. This
+  //  could possibly cause a megamorphic call site that defeats the CPU branch
+  //  predictor. The fix for this is to somehow distinguish different
+  //  functions/methods at the template level. This could be done with a counter
+  //  or a name argument of some kind.
+
+  template <typename Return, typename... Extra>
+  void operator()(
+      pybind11::module_& m, const char* name, Return (*pf)(Args...),
+      const Extra&... extra) const {
+    def_func(m, name, pf, extra...);
+  }
+
+  template <typename Func, typename... Extra>
+  void operator()(
+      pybind11::module_& m, const char* name, Func&& f,
+      const Extra&... extra) const {
+    this->template operator()(m, name, &f, extra...);
+  }
+
+  template <typename Cls, typename Return, typename ImplCls, typename... Extra>
+  void operator()(
+      pybind11::class_<Cls>& cls, const char* name,
+      Return (ImplCls::*pmf)(Args...), const Extra&... extra) const {
+    // This leaks a single pointer sized struct for each defined numba function.
+    // Repeated import could theoretically cause this to matter, but it's very
+    // unlikely.
+    def_class_method<Cls, Return>(
+        cls, name, pmf, new MemberFunction<ImplCls, Return, Args...>{pmf},
+        extra...);
```

Comment:
This should work just fine, I have no problem with it
Without understanding the full decision space, I would think a more common way to do this would be an effective Singleton, with a declaration in the MemberFunction struct and a definition in a cpp file

Reply:
I'm not totally sure what you mean. I think the static on `MemberFunction` must be in the header because we want and need a new instance of static function for every instantiation of the `MemberFunction` template. Also a singleton will not work because the function *must* be static so we can take a pointer to it and cast that pointer to `void*`. Method pointers are not castable because they have different calling conventions than normal function (on at least some compilers).

Reply:
My understanding:
In order to wrap a method for python, python wants a function pointer that accepts `(self, args...)`. In order to give it that, we're currently creating a function `MemberFunction::call` which accepts `(self, MemberFunction*, args)`, which uses the instance to remember what particular member function it's referring to. This requires a `new` instance of MemberFunction per method to be wrapped. This method generates one `MemberFunction::call` function pointer per _signature_.

Proposal:
Let's instead generate one function pointer per _method_.
```c++
#include <functional>
#include <iostream>

class Thing {
public:
  float my_method(int x, float y) { return x * y; }
};

// MemberFunction is a template on the *Signature*
template <typename Cls, typename Return, typename... Args>
struct MemberFunction {
  // call is a template on the *particular member function*
  template <Return (Cls::*member_func)(Args...)>
  static Return call(Cls *self, Args... args) {
    return (self->*member_func)(args...);
  }
};

int main() {
  std::function<float(Thing *, int, float)> function =
      MemberFunction<Thing, float, int, float>::call<&Thing::my_method>;
  Thing t;
  float z = function(&t, 6, 7.0);
  std::cout << z << "\n";
  return 0;
}
```

Only using std::function for readability, and I omitted some cvref things, but hopefully my idea is clear. I believe this can replace
`(uintptr_t)&Caller::call, (uintptr_t)caller`
in the registration line, so we don't have to make a `new` MemberFunction for each one. These will exist statically, and without duplication

Reply:
Ah! I understand what you mean now. I think I tried that and it did not work. I will try again though.

The issue I think I ran into before is that member pointers are not allowed as template arguments. As far as I know no pointers are. I don't think C++ treats them as `constexpr` because they are strictly speaking not known until dynamic link time.

Reply:
And I tried it (as you probably did) and it works. So I'm gonna need to rethink that and figure out what I failed to do correctly before. Good catch. This is exactly why I asked for someone to look at this. I clearly misunderstood what is allowed as a template parameter.

Reply:
I can't promise that the code I wrote is compliant c++, works with all compilers, or works across compilation units, as I don't really know. clang-12 seems fine with it. Happy to help!

Reply:
I pushed a rework of the code to use the approach you recommend above.